tests/unit/test_config_provider.py

```diff
@@ -0,0 +1,418 @@
+# Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+from tests import unittest
+import mock
+from nose.tools import assert_equal
+
+import botocore
+import botocore.session as session
+from botocore.configprovider import ConfigValueStore
+from botocore.configprovider import BaseProvider
+from botocore.configprovider import InstanceVarProvider
+from botocore.configprovider import EnvironmentProvider
+from botocore.configprovider import ConfigPropertyProvider
+from botocore.configprovider import ConstantProvider
+from botocore.configprovider import ChainProvider
+from botocore.configprovider import ConfigChainFactory
+
+
+class TestConfigChainFactory(unittest.TestCase):
+    def assert_chain_does_provide(self, instance_map, environ_map, config_map,
+                                  build_config_chain_args,
+                                  logical_name_to_provide, expected_value):
+        fake_session = mock.Mock(spec=session.Session)
+        fake_session.get_scoped_config.return_value = config_map
+        fake_session.instance_variables.return_value = instance_map
+        builder = ConfigChainFactory(fake_session, environ=environ_map)
+        chain = builder.build_config_chain(
+            logical_name_to_provide,
+            **build_config_chain_args
+        )
+        value = chain.provide()
+        self.assertEqual(value, expected_value)
+
+    def test_chain_builder_can_provide_instance(self):
+        self.assert_chain_does_provide(
+            instance_map={'foo': 'bar'},
+            environ_map={},
+            config_map={},
+            build_config_chain_args={},
+            logical_name_to_provide='foo',
+            expected_value='bar',
```

Comment:
So for these tests to make it even more readable, it would be nice to put the name of where it came from as the value so instead of using `bar` make the value `instance-val` to indicate it came from the instance values. You did it a little bit later on for the keys, but it would be nice to do it for the values as well.